### PR TITLE
Docker/docker-compose integration

### PR DESCRIPTION
**Pull request by santakadev**
#### What? Why?
Docker/docker-compose integration to speed up development environment setup/upgrade using Docker.

Why:
- Fast development environment setup and upgrade.
- Consistent environment.
- No virtual machine.

Why not:
- Docker integration maintenance.

#### What should we test?
- Some specs are failing and I cannot determine why.
- I introduce `DB_HOST` environment variable to specify postgresql container name. I'm not sure if this is a BC break.

#### How it works
Requirements: [Docker](https://www.docker.com/) and [docker-compose](https://docs.docker.com/compose/)

To start up the application:

- run `docker-compose up`
- Visit http://localhost:3000

Other useful commands:
```
# Load sample data
docker-compose run --rm web bundle exec rake openfoodnetwork:dev:load_sample_data

# Run tests
docker-compose run --rm web bundle exec rspec spec
docker-compose run --rm web script/karma run
docker-compose run --rm web script/karma start

# Run other commands
docker-compose run --rm web COMMAND
```

#### Pending

- [ ] Update documentation

### CODE REVIEW

**Comment by oeoeaio** [comment]
Thanks @santakadev! This is amazing! I'm just building a container on my local machine and will report back when it is done.

Just wondering: what you thing about putting this stuff in a separate repo? We could direct people who want to use docker for their dev environment there? I realise this would involve altering the setup a bit, but I think in an ideal world it would be great to keep all docker concerns completely separate from the main repo?

Instead of altering the `database.yml` file in the this repo to allow the use of a postgres container, could we overwrite that file within the container itself? I am not familiar enough with docker to know, but is it possible to ignore certain files in a volume? I am just thinking that it would be ideal if we could write our own `database.yml` to the container and ignore the one from the this repo. Does that sound workable or do you have an idea for a better approach?

So many questions!

**Comment by oeoeaio** [comment]
Hm, I did some digging and although it may be possible it doesn't sound like the approach I suggested is a standard one. Perhaps we just make the changes to `database.yml` and `environment.yml` that you have here and then move the rest into a separate repo?

**Comment by sauloperez** [comment]
> Why not:

> Docker integration maintenance.

This is exactly what concerns me the most. There have been some efforts and some people use a LXC container as specified in https://github.com/coopdevs/ofn-install/wiki/OFN-DevEnv-Quick-Start, but that works only for Linux machines. We also have an unmaintained Vagrantfile somewhere there, if I'm not mistaken. We need to outweight costs and benefits and decide if we go it. 

Regarding the config files, IMO the usual approach is to add ENV vars providing the current values as the default ones rather than using totally separate files.

**Comment by oeoeaio** [comment]
While I share @sauloperez's concerns re maintainability, I have also noticed that a lot of our time is being sucked into helping people get their dev environments set up, and this is potentially a way of mitigating that. @sauloperez could we give it a go and see if the trade-off is worth it? Scripts and other environment specific solutions are always going to be problematic. Docker feels like the right direction if we can get it up and running without too much hassle.

Also @sauloperez, can you explain in a little more detail what you meant by this?:

>the usual approach is to add ENV vars providing the current values as the default ones rather than using totally separate files.

**Comment by sauloperez** [comment]
Yep, I don't oppose and I hope the following is true!

> Docker feels like the right direction if we can get it up and running without too much hassle.

And for

> the usual approach is to add ENV vars providing the current values as the default ones rather than using totally separate files.

I mean

```ruby
development:
  adapter: postgresql
  encoding: unicode
  database: ENV.fetch('DB_DATABASE', 'open_food_network_dev')
  pool: 5
  host: ENV.fetch('DB_HOST', 'localhost')
  username: ENV.fetch('DB_USERNAME', 'ofn')
  password: ENV.fetch('DB_HOST', 'f00d')
...
```

This would allow removing the custom [config/database.travis.yml](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/config/database.travis.yml), for instance.

Let's see what others think.

**Comment by mkllnk** [comment]
Using env vars for the configuration means that we need to store these variables somewhere else. In case of Travis, yes, we can get rid of database.travis.yml and move these into .travis.yml, but in any other case we still need a file containing our env vars. Or can we put these into application.yml? That would be a good place, I reckon.

**Comment by mkllnk** [comment]
I also have a question about developing with a docker container. How do we access and edit the files with the repository. A while ago that was quite a pain to have a shared folder when I tried to work on Discourse.

**Comment by santakadev** [comment]
> Just wondering: what you think about putting this stuff in a separate repo? We could direct people who want to use docker for their dev environment there? I realise this would involve altering the setup a bit, but I think in an ideal world it would be great to keep all docker concerns completely separate from the main repo?

I think that yes, it will be better to have all Docker concerns in a separate repository, but it will take more efforts. Here is an example of what is doing Sylius, an OS e-commerce solution for PHP https://github.com/Sylius/Docker

> Instead of altering the database.yml file in the this repo to allow the use of a postgres container, could we overwrite that file within the container itself? I am not familiar enough with docker to know, but is it possible to ignore certain files in a mounted volume? I am just thinking that it would be ideal if we could write our own database.yml to the container and ignore the one from the this repo. Does that sound workable or do you have an idea for a better approach?

I like the solution proposed by @sauloperez using ENV variables with default values.

> Using env vars for the configuration means that we need to store these variables somewhere else. In case of Travis, yes, we can get rid of database.travis.yml and move these into .travis.yml, but in any other case we still need a file containing our env vars. Or can we put these into application.yml? That would be a good place, I reckon.

I'm not Ruby/Rails developer, but I think that yes, application.yml or system env.

> I also have a question about developing with a docker container. How do we access and edit the files with the repository. A while ago that was quite a pain to have a shared folder when I tried to work on Discourse.

In this case, all the changes in the working directory are in sync with the container files. This is specified in `volumes` section of the `web` service in the `docker-compose.yml` file.

**Comment by sauloperez** [comment]
> In this case, all the changes in the working directory are in sync with the container files. This is specified in volumes section of the web service in the docker-compose.yml file.

Did you experience good performance? which OS do you work with?

**Comment by santakadev** [comment]
> Did you experience good performance? which OS do you work with?

I use Ubuntu linux and the performance is good. For windows and mac, there are native and non native versions (using virtualbox) of Docker but I don't have experience with Docker in Windows/Mac.

File: config/database.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ development:
   encoding: unicode
   database: open_food_network_dev
   pool: 5
-  host: localhost
+  host: <%= ENV['DB_HOST'] %>
```
[the remark this answers is not in the document]
Comment (sauloperez) [reply config/database.yml]: I would prefer using [ENV.fetch(name, default)](https://ruby-doc.org/core-2.2.0/ENV.html#method-c-fetch), which is meant for that.

**Comment by sauloperez** [comment]
Agree with @mkllnk suggestion. I'm also afraid that it will get soon out of date but we don't oppose to someone else maintaining it.

**Comment by sauloperez** [comment]
Unless someone else is willing to take @santakadev 's work, I think we should close this PR soon. Any candidates?

**Comment by daniellemoorhead** [comment]
Closed for now. Feel free to reopen if you want to continue this work.

**Comment by pacodelaluna** [comment]
My bad, let's make a new PR, I have changed a bit the approach with Docker compose.

**Pull request closed without merging**